vnext/Microsoft.ReactNative/RedBox.cpp

```diff
@@ -11,117 +11,303 @@
-          redboxContent.FindName(L"ErrorMessageText").as<winrt::Windows::UI::Xaml::Controls::TextBlock>();
-      auto stackpanel = redboxContent.FindName(L"StackPanel").as<winrt::Windows::UI::Xaml::Controls::StackPanel>();
-      auto dismissButton = redboxContent.FindName(L"DismissButton").as<winrt::Windows::UI::Xaml::Controls::Button>();
-      auto reloadButton = redboxContent.FindName(L"ReloadButton").as<winrt::Windows::UI::Xaml::Controls::Button>();
-
-      winrt::event_token tokenDismiss;
-      tokenDismiss = dismissButton.Click([popup](
-          auto const & /*sender*/, winrt::Windows::UI::Xaml::RoutedEventArgs const & /*args*/) noexcept {
-        popup.IsOpen(false);
+          L"  >"
+          L"  <TextBlock x:Name='MethodText' FontSize='14' Foreground='White' TextWrapping='Wrap' FontFamily='Consolas'/>"
+          L"  <TextBlock x:Name='FrameText' FontSize='14' Foreground='Gray' TextWrapping='Wrap' FontFamily='Consolas'/>"
+          L"</StackPanel>";
+      auto frameContent = winrt::unbox_value<winrt::Windows::UI::Xaml::Controls::StackPanel>(
+          winrt::Windows::UI::Xaml::Markup::XamlReader::Load(xamlFrameString));
+      auto methodText = frameContent.FindName(L"MethodText").as<winrt::Windows::UI::Xaml::Controls::TextBlock>();
+      auto frameText = frameContent.FindName(L"FrameText").as<winrt::Windows::UI::Xaml::Controls::TextBlock>();
+      methodText.Text(Microsoft::Common::Unicode::Utf8ToUtf16(frame.Method));
+
+      std::stringstream stackFrameInfo;
+      stackFrameInfo << frame.File << ":" << frame.Line << ":" << frame.Column;
+      frameText.Text(Microsoft::Common::Unicode::Utf8ToUtf16(stackFrameInfo.str()));
+      m_stackPanel.Children().Append(frameContent);
+    }
+  }
+
+  winrt::Windows::UI::Xaml::Controls::StackPanel m_stackPanel{nullptr};
+  winrt::Windows::UI::Xaml::Controls::Primitives::Popup m_popup{nullptr};
+  winrt::Windows::UI::Xaml::Controls::Grid m_redboxContent{nullptr};
+  winrt::Windows::UI::Xaml::Controls::Button m_dismissButton{nullptr};
+  winrt::Windows::UI::Xaml::Controls::Button m_reloadButton{nullptr};
+  winrt::Windows::UI::Xaml::Controls::TextBlock m_errorMessageText{nullptr};
+
+  bool m_showing = false;
+  std::function<void(uint32_t)> m_onClosedCallback;
+  winrt::event_token m_tokenClosed;
+  winrt::event_token m_tokenDismiss;
+  winrt::event_token m_tokenReload;
+  winrt::event_token m_tokenSizeChanged;
+  JSExceptionInfo m_exceptionInfo;
+  Mso::WeakPtr<IReactHost> m_weakReactHost;
+};
+
+/*
+ * This class is implemented such that the methods on IRedBoxHandler are thread safe.
+ */
+struct RedBoxHandler : public std::enable_shared_from_this<RedBoxHandler>, IRedBoxHandler {
+  RedBoxHandler(
+      Mso::WeakPtr<Mso::React::IReactHost> weakReactHost,
+      std::shared_ptr<facebook::react::MessageQueueThread> uiMessageQueue) // Should uiMessageQueue be a weakRef too?
+      : m_weakReactHost(std::move(weakReactHost)), m_wkUIMessageQueue(uiMessageQueue) {}
```

Comment:
>uiMessageQueue [](start = 70, length = 14)

std::move

Reply:
Never mind: I did not realize that we assign shared_ptr to std::weak_ptr here. Then to avoid extra AddRef/Release please pass parameter as const& instead of by value.

---
In reply to: [391097181](https://github.com/microsoft/react-native-windows/pull/4290#discussion_r391097181) [](ancestors = 391097181)